pandas/core/groupby/generic.py

```diff
@@ -1143,6 +1143,9 @@ def nunique(self, dropna=True):
 
         val = self.obj._internal_get_values()
 
+        # GH 27951
+        val[isna(val)] = np.datetime64("NaT")
```

Comment:
I think the actual root of the issue is a bug in NumPy as described by @TomAugspurger where NaT values are not sorted as you'd expected

https://github.com/numpy/numpy/issues/12629

So I think this works for now but maybe add a comment about NumPy bug 12629 for reference